app/code/Magento/Quote/Model/ShippingAddressManagement.php

```diff
@@ -88,7 +88,7 @@ public function assign($cartId, \Magento\Quote\Api\Data\AddressInterface $addres
         $quote = $this->quoteRepository->getActive($cartId);
         if ($quote->isVirtual()) {
             throw new NoSuchEntityException(
-                __('The Cart includes virtual product(s) only, so a shipping address is not used.')
+                __('Shipping address is not allowed on cart: cart contains no items for shipment.')
```

Comment:
usually a patch release "2.4.1" we can't change phrases.
Since this happens in core,  I would rather introduce a new phrase in graphql only by catching NoSuchEntityException + $quote->isVirtual()